python/gym_jiminy/common/gym_jiminy/common/utils/spaces.py

```diff
@@ -263,21 +319,91 @@ def copy(data: DataNestedT) -> DataNestedT:
+                 func: Callable[[], DataNestedT],
+                 out: DataNestedT) -> None:
+        """Set a given field of a nested data structure to the value return by
+        a function with no input argument.
+
+        :param field: Field to set.
+        :param func: Function to call.
+        :param out: Nested data structure.
+        """
+        out[field] = func()
+
+    def _chain_calls(func1: Callable[[DataNestedT], None],
+                     func2: Callable[[DataNestedT], None],
+                     out: DataNestedT) -> None:
+        """Call two functions sequentially in order while passing the same
+        input argument to both of them.
+
+        :param func1: First function.
+        :param space: Second function.
+        :param out: Input argument to forward.
+        """
+        func1(out)
+        func2(out)
+
+    func = None
+    for field, subspace in dict.items(space.spaces):
+        op = partial(_setitem, field, build_clip(data[field], subspace))
+        func = op if func is None else partial(_chain_calls, func, op)
+    if func is None:
+        return OrderedDict()
+
+    # Define the chain of functions operating on a given out
+    def _clip_impl(func: Callable[[DataNestedT], None]) -> DataNestedT:
+        """Clip arbitrarily nested data structure of 'np.ndarray' specialized
+        for some pre-allocated data.
+        """
+        out = OrderedDict()
+        func(out)
+        return cast(DataNestedT, out)
+
+    return partial(_clip_impl, func)
+
+
 def clip(data: DataNestedT,
-         space_nested: gym.Space[DataNestedT],
-         check: bool = True) -> DataNestedT:
-    """Clamp value from `gym.Space` to make sure it is within bounds.
+         space: gym.Space[DataNestedT]) -> DataNestedT:
+    """Clip data from `gym.Space` to make sure it is within bounds.
 
     .. note:
         None of the leaves of the returned data structured is sharing memory
-        with the original one, even if clipping had no effect or was not
-        applicable. This alleviate the need of calling 'deepcopy' afterward.
+        with the original one, even if clipping had no effect. This alleviate
+        the need of calling 'deepcopy' afterward.
 
-    :param space: `gym.Space` on which to operate.
     :param data: Data to clip.
+    :param space: `gym.Space` on which to operate.
     """
-    if check:
-        return tree.map_structure(_clip_or_copy, data, space_nested)
-    return cast(DataNestedT, _unflatten_as(data, [
-        _clip_or_copy(value, space) for value, space in zip(
-            tree.flatten(data), tree.flatten(space_nested))]))
+    if not isinstance(space, gym.spaces.Dict):
+        return _clip_or_copy(data, space)
+
+    out = OrderedDict()
+    for field, subspace in dict.items(space.spaces):
+        out[field] = clip(data[field], subspace)
+    return out
+
+
+def contains(data: DataNestedT, space: gym.Space[DataNestedT]) -> bool:
+    """TODO: Add documentation
```

Comment:
Check if all leaves of a nested data structure are within bounds of there respective `gym.Space`.

By design, it is always `True` for all spaces but `gym.spaces.Box`, `gym.spaces.Discrete` and `gym.spaces.MultiDiscrete`.